core/modules/filter/filter.pages.inc

```diff
@@ -491,28 +491,128 @@ function filter_format_editor_dialog_save($form, &$form_state) {
-    $value = link_cleanup_url($value);
-    $form_state['values']['type'] = $type;
-    if ($type === LINK_INTERNAL && strpos($value, '/') !== 0) {
-      $value = '/' . $value;
-    }
+  // Allow empty values, since this will not result in a link tag anyway.
+  if ($value === '') {
+    form_set_value($element, $value, $form_state);
+    return;
   }
 
+  // Clean up the URL so it's more likely to pass validation.
+  $value = _filter_cleanup_url($value);
+
   form_set_value($element, $value, $form_state);
 
+  // Flag if it has any dangerous protocols.
+  $stripped_url = backdrop_strip_dangerous_protocols($value);
+  $dangerous_protocols = $stripped_url !== $value ? TRUE : FALSE;
+
   // Unlike #type = 'url' validation, we allow both relative and absolute paths.
-  if ($value !== '' && !valid_url($value, TRUE) && !valid_url($value, FALSE)) {
+  if ($dangerous_protocols || (!valid_url($value, TRUE) && !valid_url($value, FALSE))) {
     form_error($element, t('The URL %url is not valid.', array('%url' => $value)));
   }
 }
 
+/**
+ * Clean up the URL for validation.
+ *
+ * Attempts to create a URL that will pass validation: add a protocol if it
+ * looks like a domain but it's missing; or create a valid, encoded internal
+ * alias in case transliteration is disabled.
+ *
+ * @param string $url
+ *
+ * @return string
+ *   A cleaned-up URL that is either a properly formatted external URL or an
+ *   internal path.
+ *
+ * @see _filter_format_editor_link_url_validate()
+ */
+function _filter_cleanup_url($url) {
+  // Will add a default protocol if missing on an external url.
+  $url = _filter_create_valid_external_url($url);
+
+  // If it's an external URL with a protocol we can skip the rest of the
+  // clean up.
+  if (_filter_url_has_protocol($url)) {
+    return $url;
+  }
+
+  // Return anchors as is.
+  if (substr($url, 0, 1) == '#') {
+    return $url;
+  }
+
+  // Break up the components so we can build up an aliased internal path.
+  $url_parts = backdrop_parse_url($url);
+
+  return url($url_parts['path'], $url_parts);
+}
+
+/**
+ * Forms a valid external URL if possible.
+ *
+ * Trims whitespace and automatically adds an http:// to addresses that
+ * appear to be external URLs but have no specified protocol.
+ *
+ * @param string $url
+ *   The URL entered by the user.
+ *
+ * @return string
+ *   The URL with the protocol prepended if needed.
+ *
+ * @see link_cleanup_url().
+ */
+function _filter_create_valid_external_url($url) {
+  // Skip if it's already a valid external URL with a protocol.
+  if (_filter_url_has_protocol($url)) {
+    return $url;
+  }
+
+  // If it looks like a domain name but it's missing a protocol Add an
+  // an http:// or https://, depending on the site configuration.
+  $domain_match = preg_match('/^(([a-z0-9]([a-z0-9\-_]*\.){1,63})([a-z]{2,63}))/i', $url);
+  if (!empty($domain_match)) {
+    if (backdrop_is_https()) {
+      $protocol = 'https://';
```

Comment:
Thinking this over...

External urls are not necessarily also https just because the Backdrop site is. Very likely this doesn't do any harm, but it makes questionable assumptions.

Reply:
Yeah. I added this after seeing something similar elsewhere. And was also thinking of "Mixed Content" https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content but I guess that's only an issue if actually pulling in content which we're not doing here.

So I could either go back to just adding "http://" or we could make it a relative protocol: "//".

Reply:
I got the idea from https://git.drupalcode.org/project/link/-/commit/a7d2015d2208b8717acc1c6de2b884da5525f0f3. They don't explain why they checked https, but just thinking maybe it's because the resulting absolute URL might be internal to the site? It's possible that happens. Is that enough reason? But maybe just making it a relative protocol is enough.

Reply:
> maybe it's because the resulting absolute URL might be internal to the site?

I see, that way it makes sense, of course.

> But maybe just making it a relative protocol is enough.

Hehe, that seems clever. :+1: